Add support for Swift Package Manager

vauxhall opened the pull request:
Adding support for Swift Package Manager would allow your users to stop using CocoaPods or Carthage and move a solution already integrated on Xcode.

Unfortunately 2 of your headers had to be modified in order for the package to build using SPM.

I understand these changes will break things, but my knowledge handling header files is not that good, maybe somebody in your team knows how to avoid changing:

`#import <fmdb/FMDB.h>`
to
`#import <FMDB.h>`



I was able to add successfully this package as a dependency on a project using Swift 5.2

I'm sorry for spamming your CI, there is no other way to test if the configuration on the Package Manifest is going to work and if the changes I made on your headers are affecting other deployment options.

vauxhall commented:
This would be part of the issue: https://github.com/snowplow/snowplow-objc-tracker/issues/474
I actually used the example provided there as a base ground to experiment.

AlexBenny commented:
Thanks @vauxhall for your helpful contribute.

> Adding support for Swift Package Manager would allow your users to stop using CocoaPods or Carthage and move a solution already integrate on Xcode.

Totally agree. We are keen to support all the major package managers and Swift Package Manager is definitely an important one.

> I understand these changes will break things, but my knowledge handling header files is not that good, maybe somebody in your team knows how to avoid changing

Handling Objective-C headers can be troublesome sometime.
I'll give a look at the header issue in the next days.

vauxhall commented:
@AlexBenny Fixed the imports issue by validating [Swift_Package](https://github.com/apple/swift-package-manager/blob/master/Documentation/Usage.md#packaging-legacy-code) preprocessor definition, which is useful when using with already existing sources.

Now these changes are less invasive, and transparent for Carthage and CocoaPods users, and CI is green :)

AlexBenny commented:
Thanks @vauxhall , happy to hear that it passes the CI and it works properly on your app!

#### It looks good, **I'm just adding a quick list of things to fix before the release in production:**

- FMDB dependency should be 2.6 or above (excluded next major version 3.0)
- The targets section may list the testTarget too
- Add a demo app "SnowplowSwiftSPMDemo" that uses SPM as package manager

We have to add an example app that compile with SPM. We don't need to create a new one. We just need to create the app project (that uses SPM) and link (not duplicate) the code in "CommonSwiftCode" folder. Exactly like the other SnowplowSwift...Demo apps on Examples folder.
We do that because we want to test all the package managers without duplications of the source files of the demo app.

- Update travis script in order to build SnowplowSwiftSPMDemo app

As explained above. We want to be sure the tracker build correctly using all the package managers we support.


#### Here a couple of issues that (I guess) we are not able to fix now (mostly due to SPM limitations):

- The package brings with itself all the files in the repo.

Unfortunately it seems we can't exclude that our package comes with unneeded files (like Examples folder, vagrant, etc.).
SPM does a clone of the git repository so it gets all the repo each time is integrated in an app.
It seems we have to wait for an evolution in order to fix this issue.

- The package doesn't consider targets for the other platforms: watchOS, macOS and tvOS

I've played a bit adding other targets in the Package.swift but I ended up with the error "target ... has sources overlapping sources".
This seems due to restrictions of SPM that are in conflict with the way our tracker is done (some details here: https://stackoverflow.com/questions/59652708/two-almost-identical-targets-in-vapor-xcode-project). 

A pro and con of SPM is that it uses the file system as source of truth. Folders and modules are strongly connected. This obviously can be a problem for projects like this one architected with an ObjC approach.

I like the solution suggested on the StackOverflow link above but I wouldn't spend too much time changing the source files and targets structure now.
Unless you would like to restructure targets and sources (without creating breaking changes), I'm fine with letting SPM to manage the iOS target only.
We can update Package.swift for the other platforms in a second moment.

vauxhall commented:
Thanks @AlexBenny for your feedback and for reviewing options too.


> * FMDB dependency should be 2.6 or above (excluded next major version 3.0)

I cannot link it to a specific release because before version 2.7 they were not using SPM themselves yet, but I linked it to the first commit where they added they Package manifest.


> * The targets section may list the testTarget too

Added!


> * Add a demo app "SnowplowSwiftSPMDemo" that uses SPM as package manager

Will need more time for this, hopefully in the following days.


> * Update travis script in order to build SnowplowSwiftSPMDemo app

Will also need more time for this, it has been a while since I touched travis.


> Unless you would like to restructure targets and sources (without creating breaking changes), I'm fine with letting SPM to manage the iOS target only.
> We can update Package.swift for the other platforms in a second moment.

Modified the manifest so for now it only supports iOS, and we can start adding other platforms after that :)

AlexBenny commented:
Thanks @vauxhall 👍 
No problem for the travis script. I can quickly update it to work with the app once is added.

vauxhall commented:
@AlexBenny Added the Demo App as requested but it won't be able to compile using SPM until this same PR is merged, otherwise Xcode can't link it, see screenshot.


<img width="734" alt="Screenshot 2020-04-07 at 16 21 09" src="https://user-images.githubusercontent.com/4469032/78681962-b35cb100-78ed-11ea-80bd-48a619fbceff.png">


Maybe we can also do that on multiple steps?

AlexBenny commented:
> Added the Demo App as requested but it won't be able to compile using SPM until this same PR is merged, otherwise Xcode can't link it, see screenshot.

Oh I see. Have you tried to directly import the package from sources? We adopted the same approach with Cocoapods and Carthage?
Like doing:
```
  .package(path: "../Examples/SnowplowSwiftSPMDemo/")
```
I haven't checked but it seems possible from their docs: https://github.com/apple/swift-package-manager/blob/master/Documentation/PackageDescription.md

vauxhall commented:
@AlexBenny You were right, it was possible to add as a local dependency.
